pkg/providers/tf/job_runner.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2018 the Service Broker Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/gcp-service-broker/brokerapi/brokers/models"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/db_service"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/pkg/providers/tf/wrapper"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/utils"
+)
+
+const (
+	InProgress = "in progress"
+	Succeeded  = "succeeded"
+	Failed     = "failed"
+)
+
+// NewTfJobRunerFromEnv creates a new TfJobRunner with default configuration values.
+func NewTfJobRunerFromEnv() (*TfJobRunner, error) {
+	projectId, err := utils.GetDefaultProjectId()
+	if err != nil {
+		return nil, err
+	}
+
+	return &TfJobRunner{
+		ProjectId:      projectId,
+		ServiceAccount: utils.GetServiceAccountJson(),
+	}, nil
+}
+
+// TfJobRunner is responsible for executing terraform jobs in the background and
+// providing a way to log and access the state of those background tasks.
+//
+// Jobs are given an ID and a workspace to operate in, and then the TfJobRunner
+// is told which Terraform commands to execute on the given job.
+// The TfJobRunner executes those commands in the background and keeps track of
+// their state in a database table which gets updated once the task is completed.
+//
+// The TfJobRunner keeps track of the workspace and the Terraform state file so
+// subsequent commands will operate on the same structure.
+type TfJobRunner struct {
+	ProjectId      string
+	ServiceAccount string
+}
+
+// StageJob stages a job to be executed. Before the workspace is saved to the
+// database, the modules and inputs are validated by Terraform.
+func (runner *TfJobRunner) StageJob(ctx context.Context, jobId string, workspace *wrapper.TerraformWorkspace) error {
+	// Validate that TF is happy with the workspace
+	if err := workspace.Validate(); err != nil {
+		return err
+	}
+
+	workspaceString, err := workspace.Serialize()
+	if err != nil {
+		return err
+	}
+
+	deployment := &models.TerraformDeployment{
+		ID:                jobId,
+		Workspace:         workspaceString,
+		LastOperationType: "validation",
```

Comment:
This is informational, usually a Provision will be issued right after this, but if it doesn't (like if the operator killed the process) then this state will be in the DB. Nothing will be returned to the user between this call and that one, so the cloud-controller would retry with a new GUID or tell the user to retry.

I went with this approach because I always wanted workspaces to be in a valid state in the DB because it made the preconditions for Create and Destroy more consistent. In my original approach, I didn't commit anything to the DB until the job had started, but that made the Create and Destroy methods fundamentally different, when Terraform really treats them the same which felt wrong.